### PR TITLE
Use dense BitVectors in region analysis

### PR DESCRIPTION
**Pull request by ncough**
Use dense BitVectors, rather than sparse, to
implement region analysis. This reduces the
compile overhead, as allocations for the sparse
BitVectors are costly.

### CODE REVIEW

**Comment by andrewcraik** [comment]
LGTM @vijaysun-omr

File: compiler/optimizer/StructuralAnalysis.hpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ class TR_RegionAnalysis
 
    class StructInfo;
    typedef CS2::TableOf<StructInfo, TR::Allocator> InfoTable;
-   typedef TR::SparseBitVector SparseBitVector;
+   typedef TR::BitVector SparseBitVector;
```
Comment (0xdaryl) [inline compiler/optimizer/StructuralAnalysis.hpp]: This is quite subtle.  By the name alone, someone not deeply familiar with this code would likely be expecting a sparse data structure.  Perhaps a different name should be used.

**Comment by andrewcraik** [comment]
So the logic behind the change is as follows: following a conversation with @lmaisons I was given the understanding that a SparseBitVector can become more efficient than a BitVector when the number of set bits is less than 1/16. In a SparseBitVector nearly every element being set causes an allocation. The cost of the allocation is readily seen in execution profiles - especially since when running with OSR on you need to run more passes of structural analysis and one of the most expensive parts of that analysis when profiled are the allocations for the SparseBitVector. Based on the average size of methods the savings of the SparseBitVector seem small compared to the cost of running the region memory allocators and the overhead associated with them.

I am open to hearing other points of view on this one, but the significant allocation overhead for SparseBitVector made it feel like this wasn't the right tradeoff any more. @mpirvu interested to hear your thoughts as well.

File: compiler/optimizer/StructuralAnalysis.hpp (hunk 1 above)
Comment (andrewcraik) [inline compiler/optimizer/StructuralAnalysis.hpp]: I agree - a rename here would be appropriate.

File: compiler/optimizer/Structure.cpp [hunk 2]
```diff
@@ -363,8 +363,7 @@ void TR_RegionStructure::addGlobalRegisterCandidateToExits(TR_RegisterCandidate
    }
 
 
-
-static bool findCycle(TR_StructureSubGraphNode *node, TR::SparseBitVector &regionNodes, TR_BitVector &nodesSeenOnPath, TR_BitVector &nodesCleared, int32_t entryNode)
+static bool findCycle(TR_StructureSubGraphNode *node, TR::BitVector &regionNodes, TR_BitVector &nodesSeenOnPath, TR_BitVector &nodesCleared, int32_t entryNode)
```
Comment (mgaudet) [inline compiler/optimizer/Structure.cpp]: Maybe what should be done is moving the `typedef` such that we can use the `SparseBitVector` typedef in this signature -- (after it has been renamed of course. Perhaps `StructureBitVector`?

File: compiler/optimizer/Structure.hpp [hunk 3]
```diff
@@ -626,8 +626,8 @@ class TR_RegionStructure : public TR_Structure
         void reset() {_iter.SetToFirstOne();}
 
         private:
-          TR::SparseBitVector _nodes;
-          TR::SparseBitVector::Cursor _iter;
+          TR::BitVector _nodes;
+          TR::BitVector::Cursor _iter;
```
Comment (mgaudet) [inline compiler/optimizer/Structure.hpp]: Ditto to my previous comment; perhaps the `BitVector` `typedef` ought to be in Structure.hpp?

File: compiler/optimizer/StructuralAnalysis.cpp [hunk 4]
```diff
@@ -51,7 +51,7 @@ void TR_RegionAnalysis::simpleIterator (TR_Stack<int32_t>& workStack,
                                         bool doThisCheck)
    {
    SparseBitVector::Cursor cursor(vector);
-   for (cursor.SetToLastOne(); cursor.Valid(); cursor.SetToPreviousOne())
+   for (cursor.SetToFirstOne(); cursor.Valid(); cursor.SetToNextOne())
```
Comment (mgaudet) [inline compiler/optimizer/StructuralAnalysis.cpp]: Do sparse and dense bitvectors order their bits differently? Or is this a separate optimization?

Comment (andrewcraik) [inline compiler/optimizer/StructuralAnalysis.cpp]: Looking at the code there may be a small benefit to backwards iteration, but the main reason for the change is that the reverse traversal is only currently supported by the SparseBitVector::Cursor and not the BitVector::Cursor - it seems to be the only asymmetry in the API...

**Comment by mpirvu** [comment]
I vaguely recollect that SparseBitVectors were mostly useful to the static compiler.
However, to cover all bases I think we should measure the impact of this change alone on startup-time/CPU/footprint.

**Comment by lmaisons** [comment]
If the reverse iteration is beneficial, it might be worth revisiting an old patch I created to allow swapping out CS2::SparseBitVector with CS2::ABitVector using a pre-processor macro.

**Comment by vijaysun-omr** [comment]
@ncough I am assuming this needs the evaluation requested by Marius to be done before it can be merged. I am waiting for an update on this before I consider it.

**Comment by andrewcraik** [comment]
@ncough can we please work on getting the evaluation done for this one?

**Comment by mpirvu** [comment]
I performed some startup experiments using the new Liberty 17.0.0.2. 

With SCC/AOT but no -Xquickstart (the way WAS teams runs it) dense bit vectors manage to:
1) Reduce footprint by 0.26% (negligible considering CI=0.22%)
2) Increase CompCPU by 1.3% (CI=0.6%)
3) Startup time is flat.

With SCC/AOT and -Xquickstart dense bit vectors manage to:
1) Reduce footprint by 0.35% (CI=0.22%, so noise)
2) Increase CompCPU by 0.44% (CI=0.4%, so noise)
3) Improve startup time by 0.48% (CI=0.43%, so it could be noise)

I believe the expectation was that dense b it vectors would increase footprint, but improve CompCPU. If I ignore fluctuations for a moment, the results above show the exact opposite behavior. Taking fluctuations into consideration I would say the behavior is more or less the same.

**Comment by andrewcraik** [comment]
Marius evaluation shows noise - we would expect the dense BitVector to reduce the highwater mark in memory usage and, potentially, have some small improvement in compile time due to less indirection. These changes have a big impact on structure analysis which does not run much at cold. The impact will grow more for warm, hot and scorching compiles as structure is rebuilt more and the size of the structure increases.

File: compiler/optimizer/Structure.cpp (hunk 2 above)
Comment (andrewcraik) [inline compiler/optimizer/Structure.cpp]: I think that is probably a fine change, but it seems like a much bigger item than correcting the bad memory behaviour we are seeing with the current BitVector - I would be more comfortable decoupling these two issues.

File: compiler/optimizer/Structure.hpp (hunk 3 above)
Comment (andrewcraik) [inline compiler/optimizer/Structure.hpp]: Ditto to my reaction :)

**Comment by ncough** [comment]
This change has been updated to address the naming concerns. 

Testing in a benchmark that consists mostly of warm compilations shows no significant difference to the memory usage high water mark and a small reduction in compile time. This reduction is also seen in perf profiles, where the SparseBitVector allocations are immediately apparent, further amplified by OSR due to its the reliance on structure during IlGen opts. With this change applied, structural analysis and the optimizations that require it show a reduction in compile time.

**Comment by vijaysun-omr** [comment]
Looks ready to go in. Merging.

**Pull request merged**